### DISCUSSION TITLE
What should be the scope for the `@marlowe.io/wallet` package?

### DISCUSSION THREAD

**Discussion by hrajchert** [input-output-hk/marlowe-ts-sdk#106] [General]
The wallet package was the first to be implemented and initially, it was designed around the [CIP-30](https://cips.cardano.org/cips/cip30/) standard. However the scope of the package wasn't/isn't clear, and some limitations of CIP-30 popped up (for example the inability to distinguish test networks).

Being close to CIP-30, it can be tempting to solely rely on this package for any wallet information that is needed to build a DApp. However, noticing that is not currently clear how the CIP-30 will evolve, and that we also should be compatible with other standards like [CIP-45](https://github.com/cardano-foundation/CIPs/pull/395) and that other libraries like Lucid provides better/different underlying support, I would like to propose that the scope of the `@marlowe.io/wallet` package be restricted to only be an adapter layer so that higher level packages like `@marlowe.io/runtime-lifecycle` can operate.

In an upcoming PR to add support for Node.js/Deno, I'll introduce a `lucid` module inside the `wallet` package that can be used both in the browser and the backend. Users will still have access to the underlying and well-established `lucid` library to manage their wallet needs, so for example they will be able to ask for the [protocol parameters](https://deno.land/x/lucid@0.10.7/mod.ts?s=Provider#method_getProtocolParameters_0) instead of just knowing whether they are on mainnet or not (our current API).

**Comment by jinglescode** [comment]
Hey @hrajchert 

I'm glad that you are looking at decoupling `@marlowe.io/wallet` a bit from other marlowe packages like `@marlowe.io/runtime-lifecycle`.

I would highly recommend looking at integrating with [https://github.com/input-output-hk/cardano-js-sdk](https://github.com/input-output-hk/cardano-js-sdk) for all wallet related codebase, and sync with the Lace team on a common code base.

**Reply by hrajchert** [reply]
more than decoupling is coupling 😅. What I propose is to make the wallet package closely coupled to our internal needs (lifecycle package for the moment) and not be a general wallet package that should cover all the DApp's needs. For all the DApp needs you would use cardano-js-sdk/Lucid/CIP-30+extensions/CIP-45 and we would offer adaptors from those technologies to what the TS-SDK needs.

**Reply by jinglescode** [reply]
I created [MeshJS](https://meshjs.dev/), it has solid wallet integrations too, and popular onboarding developers on Cardano.

> we would offer adaptors from those technologies to what the TS-SDK needs
are you referring to include MeshJS adaptors as well as any future tootkits, that will keep you busy understanding each toolkits and building adaptors for each of them.

Likely not. Thus, highly recommend looking at integrating with https://github.com/input-output-hk/cardano-js-sdk for CIP30.

**Comment by ErickRomeroDev** [comment]
hi @hrajchert i have being using Lucid and Marlowe SDK with Nextjs & Deno. The Lucid pachage works just fine using Deno, Nextjs(page router) and in the browser. Nevertheless, when Lucid is initialized on the server with Nextjs(app-router) some initializations errors pop up. Error message below:

<node_modules/lucid-cardano/esm/src/lucid/lucid.js (60:38) @ TransactionBuilderConfigBuilder
 ⨯ TypeError: Cannot read properties of null (reading 'TransactionBuilderConfigBuilder')>

I am bringing this issue in this discussion since the wallet package is using Lucid. Nextjs is a webframework used by a vast number of developers  and they recommned using the latest App routing feature.

**Reply by nhenin** [reply]
Hi @erickromeroedda, I'll try to have a look next week, I keep you posted :-)

**Reply by ErickRomeroDev** [reply]
thanks @nhenin, any news?

**Reply by hrajchert** [reply]
Hi @Erickrs2, I've created a minimal next.js POC that uses the TS-SDK using the `create-next-app`. In the process it asked me if I wanted to use the new app router (which was recommended) and I said yes, but this example does not have routing.

Most of the issues I ran while doing the example were not particular neither to `lucid` nor the `ts-sdk`, they were `Next.js` specifics, particularly its client/server code handling. 

See if this helps you with the problem you were having, and if not, please provide a minimal example that triggers the error you mentioned

https://github.com/hrajchert/marlowe-nextjs-test

**Reply by ErickRomeroDev** [reply]
Hi @hrajchert thanks for the response. Yeap that is one of the issues that i was having Nextjs client side. I used the same approach to make it work. I
What i did not manage to make it work was to initialize the Wallet(Marlowe) with Lucid at the server side using Nextjs. What could be the approach in such a case?

The wallet and Lucid works just fine on node.js and client side, but the problem i am having is at the server side with nextjs (server component) (without using "use client")

**Reply by hrajchert** [reply]
Could you make a simple example repo like I did with the issue?

**Reply by ErickRomeroDev** [reply]
Sure, i forked the repo you send me, and add a component on the server trying to make a wallet.
https://github.com/Erickrs2/marlowe-nextjs-test.git

**Reply by hrajchert** [reply]
I've confirmed the problem, and it is related to an issue already reported in the lucid repo. Sadly the problem wasn't solved yet so there is little that I can do.

https://github.com/spacebudz/lucid/issues/174

The root cause is that lucid uses a WASM library for doing the CBOR serialization and that is not being managed well by the internal webpack configuration. The `null` that the error refers to is the WASM library.

You can comment the issue to try to increase its priority. A different path the TS-SDK can take to try to circumvent this is to use a wallet that doesn't depend on the WASM library (we wanted to do this regardless as it is a 4Mb library). I think  [@cardano-sdk/wallet](https://www.npmjs.com/package/@cardano-sdk/wallet) does not use the WASM module, but I haven't tried it yet.

If you need to have a backend wallet using the TS-SDK an alternative for the moment is to have a separate Node.js/Deno application and make the Next.js server side component to talk to that application via HTTP or file socket.

**Reply by ErickRomeroDev** [reply]
Hi Hernan, thanks for the response. 
Yeap, i was exactly doing that as an alternative (deno backend).
Looking forward for that alternative wallet, so i dont have to use a separate backend when it is not neccessary.
